thetis/tracer_eq_2d.py

```diff
@@ -268,6 +268,25 @@ def residual(self, solution, solution_old, fields, fields_old, bnd_conditions):
         return -f
 
 
+class ReactionTerm(TracerTerm):
+    r"""
+    Generic reaction term
+
+    The weak form reads
+
+    .. math::
+        F_r = \int_\Omega R \phi dx
+
+    where :math:`R` is a user defined reaction term.
+    """
+    def residual(self, solution, solution_old, fields, fields_old, bnd_conditions):
+        f = 0
+        R = fields_old.get('reaction_terms')
+        if R != 0 and R is not None:
+            f += -inner(R, self.test)*self.dx
+        return -f
+
+
```

Comment:
What is a reaction term? Not trying to be pedantic, I know within an application area it may have a well defined meaning, I'm just asking in terms of its functional meaning - which I guess is asking for mathematical rather than a physical definition. Functionally at the moment it looks like a reaction term is exactly the same as a source term. There's many source/right-hand side terms that may occur in applications and just in general I don't think we want to be adding new terms for every bit of physics if there's no functional difference wrt how Thetis handles them: the user can just provide a sum of these as the source.
Now you may say that a reaction term is a term that depends on other tracers but then we need some idea on how this coupling would work more generally - if we're just doing Picard iterations, i.e. we're treating the tracer equations one by one, then within the context of each tracer equation it's just a source term.

Reply:
I agree that it is identical to `SourceTerm`, except that the name of the field is different and it gets added as 'implicit', rather than 'source', at the end. I think the same can be achieved with `SourceTerm`, with a couple of tweaks.